### ISSUE TITLE
Implement "filter" for aliases

### ISSUE THREAD

**Issue by GermanBluefox** [ioBroker/ioBroker.js-controller#529]
Any ideas?

**Comment by Apollon77** [comment]
Waht you exactly mean with "filter"?

**Comment by GermanBluefox** [comment]
**Option 1:**
Use read/write function. If null is returned, the value will be ignored.
**Option 1.5** add filter flag for aliases and only if filter is true, null will be ignored. 
Option 2: implement filterRead and filterWrite options...

**Comment by GermanBluefox** [comment]
e.g. write: `val < 10 ? null : val`

**Comment by AlCalzone** [comment]
how about a property on the object with some predefined methods that can be used?
e.g.
```
filter: "ifGreaterThanOrEqual(10)" // equivalent to val >= 10 ? val : null
filter: "ifLessThan(9)" // equivalent to val < 9 ? val : null
...
```

Thinking about it, I would probably call that option `transform`, so we could also do stuff like:
```
transform: "multiply(5)" // equivalent to val * 5
transform: "not()" // equivalent to !val
```

**Comment by om2804** [comment]
> 
> 
> how about a property on the object with some predefined methods that can be used?
> e.g.
> 
> ```
> filter: "ifGreaterThanOrEqual(10)" // equivalent to val >= 10 ? val : null
> filter: "ifLessThan(9)" // equivalent to val < 9 ? val : null
> ...
> ```
> 
> Thinking about it, I would probably call that option `transform`, so we could also do stuff like:
> 
> ```
> transform: "multiply(5)" // equivalent to val * 5
> transform: "not()" // equivalent to !val
> ```

transform, convert, converter ... This more accurately describes the action. I vote on "converter" :)

**Comment by GermanBluefox** [comment]
We have still a question, **what the value will be returned on getState if filtered out?**

**Comment by om2804** [comment]
> 
> 
> We have still a question, **what the value will be returned on getState if filtered out?**

If it's a converter. That obligation to return the value lies with the implementation of the converter.
The converter always returns a value.

**Comment by GermanBluefox** [comment]
> > We have still a question, **what the value will be returned on getState if filtered out?**
> 
> If it's a converter. That obligation to return the value lies with the implementation of the converter.
> The converter always returns a value.

This functionality is implemented yet: https://www.iobroker.net/#en/documentation/dev/aliases.md

This issue is really about filtering functionality

**Comment by Apollon77** [comment]
> We have still a question, what the value will be returned on getState if filtered out?

Stays old value, just noch update :-) But this is nt working with the current alias concept :-(

**Comment by AlCalzone** [comment]
> This issue is really about filtering functionality

Ok, then ignore the 2nd half of my comment

**Comment by algar42** [comment]
Filtering almost always assumes we know the current and previous value. Only in this case we can implement some simple filtering such as  
FillLinear - get previous state
FillAverage - calculate mean of previous and current 
etc...
Everything we can do with only having current val is already implemented in read/write - I don't think we need to complicate things further here.

**Comment by algar42** [comment]
A little bit crazy idea... 
Can we check if one of the history adapter (history, influx, sql) is enabled at state level (btw, not only for aliases) and if so make getState/getStates return current state as usual and the last saved state from DB (or even N last saved states as an array)? 
e.g.
`{
"val":false,
"ack":false,
"ts":1573896039569,
"q":0,
"from":"system.adapter.admin.0",
"user":"system.user.admin",
"lc":1573896039569, 
"oldStates":[
{
"val":false,
"ts":1234
},
{
"val":true,
"ts":1234
}
]}`
Request to get states from DB and how much to get can be passed to getState as options object? 
This will definitely extend the functionality and will make getState more flexible.

**Comment by Apollon77** [comment]
We have getHistory function in JavaScript adapter for this.

**Comment by foxriver76** [comment]
Are there any use cases for this filter functionality? Why should I drop a whole state change of the related state?

**Comment by Apollon77** [comment]
Some real world examples would be great yes ;-) I only can think of „sorting out invalid sensor responses“ ...

On „how“ I would go with „undefined“ as value because null is an allowed value for iobroker. Undefined is kind of not allowed as state value. 
So if script returns undefined as value it is handled as it would have not happened.

**Comment by foxriver76** [comment]
For this to work, we would need to create the alias as real state (write it in db) because when value is filtered out and now someone wants to get the value, we can not reconstruct it with the alias function, because it will only give us undefined.

IMHO it would be a nice thing for us as devs to write the alias as a real state, because it would make the alias concept more consistent to the existing states concept. E. g. we would not have to convert the alias state with the read function when someone wants to get the states value. Also we would not have to explicitly emit the alias state on state change because it would be done automatically on low level if its written to the db. 

A test case for working aliases filtering could look like:
```
    it(testName + 'Should filter out undefined values', done => {
        // we subscribe to all
        context.adapter.unsubscribeForeignStates('*', () => {
            // create an state object
            context.objects.setObject(`${gid}filterOrig`, {
                common: {
                    name: 'for filter Alias',
                    type: 'number',
                    role: 'level',
                    min: 0,
                    max: 100
                },
                native: {},
                type: 'state'
            }, err => {
                expect(err).to.be.not.ok;
                // now create alias object which filters out values less than 5 on read and greater than 10 on write
                context.adapter.setForeignObject(`${gAliasID}filterAlias`, {
                    common: {
                        name: 'Test Alias',
                        type: 'number',
                        role: 'state',
                        min: 0,
                        max: 100,
                        alias: {
                            id: `${gid}filterOrig`,
                            read: 'val < 5 ? undefined : val',
                            write: 'val > 10 ? undefined : val'
                        }
                    },
                    native: {},
                    type: 'state'
                }, err => {
                    expect(err).to.be.not.ok;
                    // set a valid orig state, so read filter wont decline it
                    context.adapter.setForeignState(`${gid}filterOrig`, 11, err => {
                        expect(err).to.be.not.ok;
                        // now set a value read filter will decline
                        context.adapter.setForeignState(`${gid}filterOrig`, 3, err => {
                            expect(err).to.be.not.ok;
                            // check that alias state equals 11 and not 3 or NaN
                            context.adapter.getForeignState(`${gAliasID}filterAlias`, (err, state) => {
                                expect(state.val).to.be.equal(11);
                                // now we set the alias to a filtered out value -> 20
                                context.adapter.setForeignState(`${gAliasID}filterAlias`, 20, err => {
                                    expect(err).to.be.not.ok;
                                    // finally check that value has been filtered out -> orig state still has to be 3
                                    context.adapter.getForeignState(`${gAliasID}filterOrig`, (err, state) => {
                                        expect(state.val).to.be.equal(3);
                                        done();
                                    });
                                });
                            });
                        });
                    });
                });
            });
        });
    });
```

**Comment by Apollon77** [comment]
One of the topics is that we would need to hold "n" real alias states in sync with the real value ... in any case - also in edge cases. We currently do not have "atomic" logic, everything is async. Also we would need to define "who" is doing this ... The writing instance/adapter/cli would need to make sure!?

**Comment by foxriver76** [comment]
The only possibility I see is to enhance the existing state object with its current alias state. Otherwise we cannot stay 100 % in sync, and we simply need to store the alias somwhere if we cannot get it's value by transforming the current state. Also `ts` and `lc`, `ack` and co have to stay at the values of the last not filtered value.

**Comment by Apollon77** [comment]
You are completely right, I also see this as the only way. Beside other options where we would limit the usecases. But in fact exactly this was what @GermanBluefox wanted to avoid because in fact you never really hold that in sync in all cases

**Comment by foxriver76** [comment]
So here we should decide if we want aliases to be really stored instead of being just a transform function as it is currently. If we keep aliases as transformers, then we can close it as it is technically not possible.

**Comment by Diginix** [comment]
Is there a way to have no value change of alias for specific values of source e.g. `null`
An empty string value with this rule isn't a real solution `val == null ? '' : val`

**Comment by foxriver76** [comment]
> Is there a way to have no value change of alias for specific values of source e.g. `null` An empty string value with this rule isn't a real solution `val == null ? '' : val`

No, this is what this ticket is about
